### DISCUSSION TITLE
MS Live Share extension: authentication problem

### DISCUSSION THREAD

**Discussion by gsmachado** [coder/code-server#2530] [General]
Hello everyone,

So, I tried to use `MS-vsliveshare.vsliveshare-pack` with `code-server` and I will describe a bit my journey here, in a summarized way:

I've made a `Dockerfile` with the steps below.

First, I uncompressed the `MS-vsliveshare.vsliveshare-1.0.3375.vsix` to the `/config/extensions` folder, as it should.

```
root@60e3d5be7552:~/extensions# ls -la
total 48
drwxr-xr-x 1 neo  neo  4096 Dec 29 14:53 .
drwxr-xr-x 1 neo  neo  4096 Dec 29 14:49 ..
drwxr-xr-x 8 neo neo 4096 Dec 29 14:54 MS-vsliveshare.vsliveshare-1.0.3375
drwxr-xr-x 1 neo  neo  4096 Dec 26 13:16 redhat.java-0.61.0
drwxr-xr-x 1 neo  neo  4096 Dec 26 13:16 vscjava.vscode-java-debug-0.29.0
drwxr-xr-x 1 neo  neo  4096 Dec 26 13:16 vscjava.vscode-java-dependency-0.14.0
drwxr-xr-x 1 neo  neo  4096 Dec 26 13:16 vscjava.vscode-java-test-0.25.0
drwxr-xr-x 1 neo  neo  4096 Dec 26 13:16 vscjava.vscode-maven-0.21.2
root@60e3d5be7552:~/extensions#
```

Second, I started `code-server` version `3.8.0` with the following command:

```
/usr/local/bin/code-server \
	--bind-addr 0.0.0.0:8443 \
	--user-data-dir /config/data \
	--extensions-dir /config/extensions \
	--disable-update-check \
	--enable-proposed-api ms-vsliveshare.vsliveshare \
	--auth "${AUTH}" \
	"${PROXY_DOMAIN_ARG}" \
	/config/workspace
```

The extension was successfully loaded, and I could get to the page where I could share something:

![image](https://user-images.githubusercontent.com/2124276/103292762-c15be000-49ee-11eb-9566-b34c90c62087.png)

When I click on "Share now", it goes to this screen to start the auth process:

![image](https://user-images.githubusercontent.com/2124276/103292811-dafd2780-49ee-11eb-9e20-73989bf01a2c.png)

Then I click "Sign in with Microsoft" and go through the authentication process, as it should normally happen. Once the sign-in process is successful, the browser is redirected to the following URL:

```
http://localhost:8080/callback?code=X.XX_XXX.XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX&state=41898cdb80097fe8f494547e43b2e5c
```

and the following screen appears:

![image](https://user-images.githubusercontent.com/2124276/103292991-447d3600-49ef-11eb-9fbc-7fd292134d00.png)

So... I investigated a bit the `code-server` code, and I found out the following:

https://github.com/cdr/code-server/blob/c4610f7829701aadb045d450013b84491c30580d/src/node/routes/vscode.ts#L125-L165

Where, [in these lines](https://github.com/cdr/code-server/blob/c4610f7829701aadb045d450013b84491c30580d/src/node/routes/vscode.ts#L125-L127), they say the `/callback` is implemented according to the `vscode` implementation:

https://github.com/cdr/code-server/blob/c4610f7829701aadb045d450013b84491c30580d/lib/vscode/resources/web/code-web.js#L419-L451

However, I've got to the conclusion that:

- Either `MS-vsliveshare.vsliveshare-1.0.3375.vsix` is not compatible to the `vscode` packed by the `code-server` version `3.8.0`;
- Or, `MS-vsliveshare.vsliveshare-1.0.3375.vsix` requires some other functions which are not open-source;
- Or, my `code-server` set-up has some kind of misconfiguration that would accept external OAuth callbacks;
- Or, the OAuth server should be externalized from `code-server` and, upon success, the `code-server` callback API should be called with some sort of token/id;

Can someone enlighten me here? I would like to discuss what's missing. Maybe I could even contribute to make this work.

**Comment by code-asher** [comment]
Huh...with GitHub authentication it redirects to an external oauth
server/relay with a callback URI (which includes the ID), the relay
redirects to GitHub, GitHub redirects back to the relay, then the relay
redirects to the callback it was provided in the first redirect after
adding the code and state to it.

It's definitely possible Microsoft authentication also uses the relay
but something doesn't go quite right there for some reason, like not
giving the ID in the first place for some reason or somehow redirecting
directly back to code-server without hitting the relay on the way back?

What does the redirect chain look like when you authenticate with
Microsoft?

**Comment by klausmcm** [comment]
What if you use the code entry approach instead?

https://docs.microsoft.com/en-us/visualstudio/liveshare/use/vscode#sign-in-using-a-user-code

Code can also be obtained from https://insiders.liveshare.vsengsaas.visualstudio.com/auth/login
